### DISCUSSION TITLE
Programmatically un-minimize a window

### DISCUSSION THREAD

**Discussion by fiddler55** [fltk-rs/fltk-rs#1605] [Q&A]
My application window can create a secondary window. When that window is hidden I can reshow it by:
`wind.show();`
However if that window is minimized, calling .show() does not un-minimize it. Is there a way to un-minimize and show a minimized window without requiring the user to go to the task bar?

**Comment by MoAlyousef** [comment]
Hi
According to the [FLTK docs](https://fltk.gitlab.io/fltk/classFl__Window.html#a473a9728b4275ce16347e74faafb3cde), `show` should work to uniminimize a window. 
The following program works for me on Windows whether the window is iconized programatically or manually. The timer calls show and the window shows.
Unfortunately I currently don't have a native linux platform.
```rust
use fltk::{prelude::*, enums::*, *};

fn main() {
    let a = app::App::default();
    let mut w = window::Window::default().with_size(400, 300);
    let mut b = button::Button::new(160, 100, 80, 30, "Click");
    w.end();
    w.show();

    b.set_callback({
        let mut w = w.clone();
        move |_| {
            w.iconize();
        }
    });

    app::add_timeout(5.0, move || {
        println!("Showing");
        w.show();
    });

    a.run().unwrap();
}
```
What platform are you not seeing this behavior?

**Comment by fiddler55** [comment]
Tried it on both Windows and Linux (Linux mint 21.2). It does indeed work on Windows, but has no effect on Linux - the window stays on the task bar.

My application has 2 help windows. If either help window is closed (hidden), they can be brought back by a call to .show() from the main window. But if the user minimizes either window, it appears there is no way to programmatically re-show the window if running on Linux.

If this remains the case, perhaps there is an alternative solution. Is there an easy way to remove the minimize button on the title bar so that the user can only close (hide) the help window when it is no longer needed? There is no real need to minimize it. (Of course it would have to work on both Windows and Linux...)

**Reply by MoAlyousef** [reply]
Searching online it appears there's no portable way of achieving this on linux due to the variability of window managers, display servers and decorators. 
You can create your own custom title bar which would only have a close button on it:
```rust
use fltk::{app, button::Button, frame::Frame, prelude::*, window::Window};

pub mod titlebar {
    use fltk::{prelude::*, enums::*, *};
    
    pub struct CustomTitlebar {
        close: button::Button,
    }
    
    impl CustomTitlebar {
        pub fn new(title: &str) -> Self {
            let parent = group::Group::try_current().unwrap();
            let width = parent.w();
            let mut win: window::Window = window::Window::from_dyn_widget(&parent).unwrap();
            win.set_border(false);
            let mut row = group::Flex::new(0, 0, width, 30, None).row();
            row.set_color(Color::Light1);
            row.set_frame(FrameType::FlatBox);
            let mut title = frame::Frame::default().with_label(title);
            title.set_label_font(Font::HelveticaBold);
            let mut close = button::Button::default().with_label("@1+");
            close.visible_focus(false);
            close.set_color(Color::Light1);
            close.set_frame(FrameType::FlatBox);
            close.set_callback({
                let mut win = win.clone();
                move |_| win.hide()
            });
            row.fixed(&close, 30);
            row.end();
            title.handle({
                let mut x = 0;
                let mut y = 0;
                let mut win = win.clone();
                move |_, event| match event {
                    Event::Push => {
                        let coords = app::event_coords();
                        x = coords.0;
                        y = coords.1;
                        true
                    }
                    Event::Drag => {
                        win.set_pos(app::event_x_root() - x, app::event_y_root() - y);
                        true
                    }
                    _ => false,
                }
            });
            Self {
                close
            }
        }

        pub fn close_button(&self) -> button::Button {
            self.close.clone()
        }
    }
}



fn main() {
    let app = app::App::default();
    let mut wind = Window::default().with_size(400, 300);
    let _title_bar = titlebar::CustomTitlebar::new("MyApp");
    let mut frame = Frame::default().with_size(200, 100).center_of(&wind);
    let mut but = Button::new(160, 210, 80, 40, "Click me!");
    wind.end();
    wind.show();

    but.set_callback(move |_| frame.set_label("Hello world"));

    app.run().unwrap();
}
```
This hides the default system provided window decorator, and creates a custom title bar:
![image](https://github.com/user-attachments/assets/5514794e-9c5e-4ca1-9512-c0575d07d39c)

**Reply by fiddler55** [reply]
Thanks for the code sample showing how to do the "alternative solution". I was able to run the code directly and I got the the image you attached, and it worked correctly. However when I integrated the "CustomTitlebar" into the help windows of my application, the compiler complains that the "close" field is never read and the "close_button" method is never used. Not sure why - it appears what I'm adding to the windows has some strange effect. But the help windows are correctly non-minimizable and can only be closed (and later reshown if desired). I'll continue to investigate, but for now a "#[allow(dead_code)]" turns off the warnings. Let me know if you have any ideas.

One of my help windows is meant to be resizable, but this has stopped working with the new title bar. Is there a way to make it resizable again?

Also the title bar styling differs from the main window. This is OK (they are help windows), but it would be better if they appeared the same as the main window. Is there a way to use the regular title bar and turn off the minimize (and maximize) buttons?

BTW I'm basically converting a Java program to Rust. The underlying Java software correctly restores an iconized window in both Windows and Linux when the following is called:

```
    if ( wind.getState() == Frame.ICONIFIED ) wind.setState(Frame.NORMAL);
    wind.setVisible(true);
    wind.toFront();

```
Don't know if something similar is available on fltk. On Linux I tried .maximize() and .un_maximize() but they had no affect on an iconized window.

**Reply by MoAlyousef** [reply]
Regarding the warning, it's because the close button is exposed to the developer if they needed to modify the callback or style. It doesn't need to be though. And the method and member button can be removed.

The styling here is custom, so if you wanted to target the styles of each OS, you might need to do that specify that manually.

Undecorated windows can't be resized with the window manage, you can handle resizability of the window manually, but that would just add so much code. 

You might want to ask in the [fltk general](https://groups.google.com/g/fltkgeneral) forum, regarding the fact that a minimized/iconized window doesn't respond to `show` calls on linux. You also have the option of asking in the fltk discussions page:
https://github.com/fltk/fltk/discussions
